homeassistant/components/switch/shelly.py

```diff
@@ -0,0 +1,111 @@
+"""
+Support for The Shelly Wifi switch.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/switch.shelly/
+"""
+
+import logging
+import requests
+import voluptuous as vol
+
+from homeassistant.components.switch import SwitchDevice, PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_HOST, CONF_NAME, CONF_PATH,
+    CONF_USERNAME, CONF_PASSWORD, CONF_SWITCHES)
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_PATH = "/relay/0"
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_PATH, default=DEFAULT_PATH): cv.string,
+    vol.Optional(CONF_USERNAME): cv.string,
+    vol.Optional(CONF_PASSWORD): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SWITCHES): vol.Schema({cv.slug: SWITCH_SCHEMA}),
+})
+
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """Set up Shelly Wifi switches."""
+    devices = []
+
+    for dev_name, properties in config[CONF_SWITCHES].items():
+        devices.append(
+            ShellySwitch(
+                hass,
+                properties.get(CONF_NAME, dev_name),
+                properties.get(CONF_HOST, None),
+                properties.get(CONF_PATH, DEFAULT_PATH),
+                properties.get(CONF_USERNAME, None),
+                properties.get(CONF_PASSWORD)))
+
+    add_devices_callback(devices)
+
+
+class ShellySwitch(SwitchDevice):
+    """Representation of a Shelly Wifi switch."""
+
+    def __init__(self, hass, name, host, path, user, passwd):
+        """Initialize the device."""
+        self._hass = hass
+        self._name = name
+        self._state = False
+        self._url = 'http://{}{}'.format(host, path)
+        if user is not None:
+            self._auth = (user, passwd)
+        else:
+            self._auth = None
+
+    def _switch(self, newstate):
+        """Switch on or off."""
+        _LOGGER.info("Switching to state: %s", newstate)
+
+        try:
+            req = requests.get('{}?turn={}'.format(self._url, newstate),
```

Comment:
Move this to an external package

Reply:
Can you please be more specific? What should I move to an external package? I'm new to this, so your input will be quite helpful.

Reply:
All device specific communication and error handling should be handled by an external package that is imported from pypi. For example you could create a shelly device class in this external package that hass can instantiate and call methods to set state to on or off.

Reply:
Read through the devdocs, its all listed there 
https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services